### PR TITLE
Change search state

### PR DESCRIPTION
**Pull request by valeskus**

### CODE REVIEW

File: src/stores/search/searchReducer.ts [hunk 1]
```diff
@@ -12,32 +11,69 @@ export interface SearchState {
   offset: number;
 }
 
+export interface SearchState {
+  searchTerm: string;
+  pendingOptions: Options;
```
Comment (MAKARD) [inline src/stores/search/searchReducer.ts]: I'd suggest having this part of the state optional and setting it to undefined upon resetting

File: src/screens/RecipesList/useRecipesListController.ts [hunk 2]
```diff
@@ -32,26 +32,26 @@ export const useRecipeListController = () => {
...
+  }, [searchOptions.pendingOptions.filter]);
 
   useEffect(() => {
-    if (!searchOptions.offset) {
+    if (!searchOptions.pendingOptions.offset) {
       return;
     }
 
-    getRecipes(searchOptions);
+    getRecipes({ searchTerm: searchOptions.searchTerm, ...searchOptions.pendingOptions });
```
Comment (MAKARD) [inline src/screens/RecipesList/useRecipesListController.ts]: searchOptions.pendingOptions can contain searchTerm.
You wrote the code in a way that "searchTerm: searchOptions.searchTerm" will be overwritten by "...searchOptions.pendingOptions"

File: src/screens/RecipesList/useRecipesListController.ts [hunk 3]
```diff
@@ -32,26 +32,26 @@ export const useRecipeListController = () => {
...
+  }, [searchOptions.pendingOptions.sort, searchOptions.searchTerm]);
 
   useEffect(() => {
     if (!recipes.length) {
       return;
     }
 
-    updateFilter(searchOptions);
-  }, [searchOptions.filter]);
+    updateFilter({ searchTerm: searchOptions.searchTerm, ...searchOptions.pendingOptions });
```
Comment (MAKARD) [inline src/screens/RecipesList/useRecipesListController.ts]: earchOptions.pendingOptions can contain searchTerm.
You wrote the code in a way that "searchTerm: searchOptions.searchTerm" will be overwritten by "...searchOptions.pendingOptions"

File: src/screens/RecipesList/useRecipesListController.ts [hunk 4]
```diff
@@ -32,26 +32,26 @@ export const useRecipeListController = () => {
   useEffect(() => {
     setLoading(true);
 
-    Promise.all([getRecipes(searchOptions), getCardType()])
+    Promise.all([getRecipes({ searchTerm: searchOptions.searchTerm, ...searchOptions.pendingOptions }), getCardType()])
```
Comment (MAKARD) [inline src/screens/RecipesList/useRecipesListController.ts]: earchOptions.pendingOptions can contain searchTerm.
You wrote the code in a way that "searchTerm: searchOptions.searchTerm" will be overwritten by "...searchOptions.pendingOptions"

Comment (MAKARD) [inline src/screens/RecipesList/useRecipesListController.ts]: I believe it is worth having a dedicated function/memorized object for storing the merged options

File: src/stores/search/searchReducer.ts [hunk 5]
```diff
@@ -12,32 +11,69 @@ export interface SearchState {
...
+        pendingOptions: {
+          filter: filter || state.pendingOptions.filter,
+          sort: sort || state.pendingOptions.sort,
+          offset: offset ?? state.pendingOptions.offset,
+        },
+      };
+    }
+
+    case SearchActions.SET_OPTIONS: {
+      const { searchTerm, sort, filter, offset } = state.pendingOptions as SearchOptionsModel;
```
Comment (MAKARD) [inline src/stores/search/searchReducer.ts]: why do you need "as" here?

File: src/stores/search/searchReducer.ts [hunk 6]
```diff
@@ -12,32 +11,69 @@ export interface SearchState {
...
+
+      return {
+        ...state,
```
Comment (MAKARD) [inline src/stores/search/searchReducer.ts:49]: don't you need to reset pending options here?

**Review by MAKARD — CHANGES_REQUESTED** [review]

File: src/screens/RecipesList/useRecipesListController.ts (hunk 2 above)
Comment (valeskus) [reply src/screens/RecipesList/useRecipesListController.ts]: searchOptions.pendingOptions not contain searchTerm....

File: src/stores/search/searchReducer.ts [hunk 7]
```diff
@@ -12,32 +11,64 @@ export interface SearchState {
...
     }
 
     case SearchActions.RESET_OPTIONS: {
       return initialState;
     }
 
+    case SearchActions.RESET_PENDING_OPTIONS: {
+      return {
+        ...state,
+        pendingOptions: initialState.pendingOptions,
```
Comment (MAKARD) [inline src/stores/search/searchReducer.ts]: since it's an optional part of state, you need to set "undefined" here

**Review by MAKARD — CHANGES_REQUESTED** [review]

File: src/stores/search/searchReducer.ts [hunk 8]
```diff
@@ -2,35 +2,60 @@ import * as Redux from 'redux';
...
-    case SearchActions.SET_OPTIONS: {
+    case SearchActions.SET_PENDING_OPTIONS: {
       const { searchTerm, sort, filter, offset } = action.payload as SearchOptionsModel;
 
       return {
         ...state,
         searchTerm: searchTerm || state.searchTerm,
-        filter: filter || state.filter,
-        sort: sort || state.sort,
         offset: offset ?? state.offset,
```
Comment (MAKARD) [inline src/stores/search/searchReducer.ts]: search term and offset are not "pendingOptions"

**Review by MAKARD — CHANGES_REQUESTED** [review]

File: src/stores/search/searchReducer.ts (hunk 8 above)
Comment (MAKARD) [reply src/stores/search/searchReducer.ts]: up

File: src/screens/RecipesList/components/RecipesListControls/RecipesListControls.tsx [hunk 9]
```diff
@@ -10,7 +10,7 @@ interface Props {
   onCardTypeChange: (type: 'grid' | 'linear') => void;
   gridType: 'grid' | 'linear';
   isFilterActive: boolean;
-  activeSort: string;
+  activeSort: string | undefined;
```
Comment (MAKARD) [inline src/screens/RecipesList/components/RecipesListControls/RecipesListControls.tsx:13]: activeSort?:

File: src/screens/RecipesList/components/RecipesListControls/useRecipesListControlsController.ts [hunk 10]
```diff
@@ -6,7 +6,7 @@ import { Icons } from '@UI/Icons';
 export const useRecipesListControlsController = (
   onCardTypeChange: (type: 'grid' | 'linear') => void,
   gridType: 'grid' | 'linear',
-  activeSort: string
+  activeSort: string |undefined
```
Comment (MAKARD) [inline src/screens/RecipesList/components/RecipesListControls/useRecipesListControlsController.ts]: activeSort?:

**Review by MAKARD — APPROVED** [review]

File: src/screens/RecipesList/components/RecipesListControls/RecipesListControls.tsx (hunk 9 above)
Comment (MAKARD) [inline src/screens/RecipesList/components/RecipesListControls/RecipesListControls.tsx]: activeSort?:

**Review by MAKARD — CHANGES_REQUESTED** [review]

**Review by MAKARD — APPROVED** [review]

**Pull request merged**